File: tools/go/agentos/task.go
// AgentOS Go SDK Task
// Version: 1.0.0.5
// Last updated: 2026-03-21

package agentos

import (
	"context"
	"errors"
	"fmt"
	"time"
)

// TaskStatus represents the status of a task
type TaskStatus string

const (
	TaskStatusPending   TaskStatus = "pending"
	TaskStatusRunning   TaskStatus = "running"
	TaskStatusCompleted TaskStatus = "completed"
	TaskStatusFailed    TaskStatus = "failed"
	TaskStatusCancelled TaskStatus = "cancelled"
)

// Task represents a task in the AgentOS system
type Task struct {
	client *Client
	ID     string
}

// NewTask creates a new Task object
func NewTask(client *Client, taskID string) *Task {
	return &Task{
		client: client,
		ID:     taskID,
	}
}

// Query queries the task status
func (t *Task) Query(ctx context.Context) (TaskStatus, error) {
	path := fmt.Sprintf("/api/tasks/%s", t.ID)
	
	response, err := t.client.request(ctx, "GET", path, nil)
	if err != nil {
		return "", err
	}
	
	statusStr, ok := response["status"].(string)
	if !ok {
		return "", errors.New("invalid response: missing status")
	}
	
	return TaskStatus(statusStr), nil
}

// Wait waits for the task to complete
func (t *Task) Wait(ctx context.Context, timeout time.Duration) (TaskResult, error) {
	start := time.Now()
	for {
		status, err := t.Query(ctx)
		if err != nil {
			return TaskResult{}, err
		}
		
		if status == TaskStatusCompleted || status == TaskStatusFailed || status == TaskStatusCancelled {
			path := fmt.Sprintf("/api/tasks/%s", t.ID)
			response, err := t.client.request(ctx, "GET", path, nil)
			if err != nil {
				return TaskResult{}, err
			}
			
			output, _ := response["output"].(string)
			error, _ := response["error"].(string)
			
			return TaskResult{
				ID:     t.ID,
				Status: status,
				Output: output,
				Error:  error,
			}, nil
		}
		
		if timeout > 0 && time.Since(start) > timeout {
			return TaskResult{}, fmt.Errorf("task timed out after %v", timeout)
		}
		
		time.Sleep(500 * time.Millisecond)
	}
}

// Cancel cancels the task
func (t *Task) Cancel(ctx context.Context) (bool, error) {
	path := fmt.Sprintf("/api/tasks/%s/cancel", t.ID)
	
	response, err := t.client.request(ctx, "POST", path, nil)
	if err != nil {
		return false, err
	}
	
	success, ok := response["success"].(bool)
	if !ok {
		return false, errors.New("invalid response: missing success")
	}
	
	return success, nil
}

// TaskResult represents the result of a task
type TaskResult struct {
	ID     string
	Status TaskStatus
	Output string
	Error  string
}

File: tools/rust/src/task.rs
// AgentOS Rust SDK Task
// Version: 1.0.0.5
// Last updated: 2026-03-21

use serde_json::{json, Value};
use std::time::{Duration, Instant};

use crate::{AgentOSError, Client};

/// Task status
enum TaskStatus {
    Pending,
    Running,
    Completed,
    Failed,
    Cancelled,
}

impl TaskStatus {
    fn from_str(s: &str) -> Option<Self> {
        match s {
            "pending" => Some(TaskStatus::Pending),
            "running" => Some(TaskStatus::Running),
            "completed" => Some(TaskStatus::Completed),
            "failed" => Some(TaskStatus::Failed),
            "cancelled" => Some(TaskStatus::Cancelled),
            _ => None,
        }
    }
}

/// Task result
#[derive(Debug)]
pub struct TaskResult {
    pub id: String,
    pub status: String,
    pub output: Option<String>,
    pub error: Option<String>,
}

/// Task
type Result<T> = std::result::Result<T, AgentOSError>;

/// Task
type JsonObject = serde_json::Map<String, Value>;

/// Task
#[derive(Debug, Clone)]
pub struct Task {
    client: Client,
    id: String,
}

impl Task {
    /// Create a new Task object
    pub fn new(client: Client, id: String) -> Self {
        Task {
            client,
            id,
        }
    }
    
    /// Query the task status
    pub async fn query(&self) -> Result<String> {
        let path = format!("/api/tasks/{}", self.id);
        let response = self.client.request("GET", &path, None).await?;
        
        let status = response.get("status")
            .and_then(|v| v.as_str())
            .ok_or_else(|| AgentOSError::InvalidResponse("Missing status".to_string()))?;
        
        Ok(status.to_string())
    }
    
    /// Wait for the task to complete
    pub async fn wait(&self, timeout: Option<Duration>) -> Result<TaskResult> {
        let start = Instant::now();
        
        loop {
            let status_str = self.query().await?;
            
            match TaskStatus::from_str(&status_str) {
                Some(TaskStatus::Completed) | Some(TaskStatus::Failed) | Some(TaskStatus::Cancelled) => {
                    let path = format!("/api/tasks/{}", self.id);
                    let response = self.client.request("GET", &path, None).await?;
                    
                    let output = response.get("output").and_then(|v| v.as_str()).map(|s| s.to_string());
                    let error = response.get("error").and_then(|v| v.as_str()).map(|s| s.to_string());
                    
                    return Ok(TaskResult {
                        id: self.id.clone(),
                        status: status_str,
                        output,
                        error,
                    });
                }
                _ => {
                    if let Some(timeout) = timeout {
                        if start.elapsed() > timeout {
                            return Err(AgentOSError::Timeout("Task timed out".to_string()));
                        }
                    }
                    
                    tokio::time::sleep(Duration::from_millis(500)).await;
                }
            }
        }
    }
    
    /// Cancel the task
    pub async fn cancel(&self) -> Result<bool> {
        let path = format!("/api/tasks/{}/cancel", self.id);
        let response = self.client.request("POST", &path, None).await?;
        
        let success = response.get("success")
            .and_then(|v| v.as_bool())
            .ok_or_else(|| AgentOSError::InvalidResponse("Missing success".to_string()))?;
        
        Ok(success)
    }
}
